### PR TITLE
multibuy - list multiple licences

### PR DESCRIPTION
**Pull request by ScottDormand96**
https://eaflood.atlassian.net/browse/IWTF-3007

On licence information page display all licences purchased

### CODE REVIEW

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js [hunk 1]
```diff
@@ -22,73 +26,188 @@ jest.mock('../../../../constants', () => ({
...
+
+      it('returns expected data', async () => {
+        const data = await getData(getMockRequest(status, transaction))
+        expect(data).toMatchSnapshot()
+      })
+    })
+
+    describe('throws a Boom forbidden error', () => {
+      it('if status agreed flag is not set', async () => {
+        const status = () => ({})
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js:190]: As a continuation of my point about `status` and `transaction` above - you never actually set a different value for transaction. So there's no need to have it as a parameter in the `getMockRequest` function

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk [hunk 2]
```diff
@@ -47,21 +21,52 @@
...
+                    {% set lengthText = mssgs.licence_type_12m %}
+                {% endif %}
 
-            <h2 class="govuk-heading-l">{{ data.permission.referenceNumber }}</h2>
+                {% if licence.ageConcession %}
+                    {% if licence.ageConcession.type == mssgs.senior %}
+                        {% set ageConcessionText = mssgs.age_senior_concession %}
+                    {% elseif licence.ageConcession.type == mssgs.age_junior %}
+                        {% set ageConcessionText = mssgs.age_junior_concession %}
+                    {% endif%}
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk:50]: Similarly, `ageConcessionText` should be setup in `getData`

**Review by jaucourt — COMMENTED** [review]
Saving my comments, may be actioned later

Resolved all these comments so that this review is a bit clearer to understand.

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js [hunk 3]
```diff
@@ -26,149 +26,156 @@ jest.mock('../../../../constants.js', () => ({
...
+      }),
+      i18n: {
+        getCatalog: () => catalog
+      }
     })
 
-    it('return value of licenceTypeDisplay is used for licenceTypeStr', async () => {
-      const licenceDisplay = Symbol('licence display')
-      licenceTypeDisplay.mockReturnValueOnce(licenceDisplay)
+    const mockRequest = getMockRequest(getMockStatus(), getMockTransaction())
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js:85]: As I've said before, don't reuse mocks. Generate a new one for every test.

Also if the default of status and transaction get functions in the mock request generator should be getMockStatus and getMockTransaction, make them that, don't give them defaults that never get used.

Please check all through your test file for doing this - I can see other places where you're defining variables used in the test outside of the test.

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js [hunk 4]
```diff
@@ -26,149 +26,156 @@ jest.mock('../../../../constants.js', () => ({
...
+    const mssgs = mockRequest.i18n.getCatalog()
 
-      const { licenceTypeStr } = await getData(createMockRequest())
+    beforeEach(() => jest.clearAllMocks())
 
-      expect(licenceTypeStr).toBe(licenceDisplay)
+    it('licenceTypeDisplay is called with permissions and mssgs', async () => {
+      await getData(mockRequest)
+      expect(licenceTypeDisplay).toHaveBeenCalledWith(getMockPermission(), mssgs)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js:92]: You're getting a second instance of an object and comparing it with the first. Don't do this, create the permission, pass it to the mock request generator and then check that licenceTypeDisplay has been called with the same object.

**Review by jaucourt — CHANGES_REQUESTED** [review]
Some common issues coming up here. Would you mind going through your code, particularly the test, checking for stuff I've told you about before - I think some of the code in here is a bit old so you've not applied changes we've discussed in the meantime.

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js [hunk 5]
```diff
@@ -26,149 +25,166 @@ jest.mock('../../../../constants.js', () => ({
...
-    {
-      type: CONCESSION.SENIOR,
-      proof: {
-        type: CONCESSION_PROOF.none
-      }
-    }
-  ]
-})
+    describe('throws a Boom forbidden error', () => {
+      it('boom is called', async () => {
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js:165]: We just need a bit better description here. Why is boom called? What is different about the data that means an error is thrown?

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js [hunk 6]
```diff
@@ -26,149 +25,166 @@ jest.mock('../../../../constants.js', () => ({
...
+        },
+        {
+          status: { [COMPLETION_STATUS.agreed]: 'agreed' },
+          boomError: 'Attempt to access the licence information handler with no posted flag set',
+          description: 'status posted flag is not set'
+        },
+        {
+          status: { [COMPLETION_STATUS.agreed]: 'agreed', [COMPLETION_STATUS.posted]: 'posted' },
+          boomError: 'Attempt to access the licence information handler with no finalised flag set',
+          description: 'status finalised flag is not set'
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/__tests__/route.spec.js:183]: great descriptions!

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk [hunk 7]
```diff
@@ -39,21 +21,44 @@
...
+                {% endif %}
+                {% if licence.ageConcession %}
+                    {% set licenseDetailsSummaryRows = (licenseDetailsSummaryRows.push({ key: { text: mssgs.age_concession }, value: { text: licence.ageConcession } }), licenseDetailsSummaryRows) %}
+                {% endif %}
+
+                <h2 class="govuk-heading-l">{{ licence.referenceNumber }}</h2>
+
+                {{ govukSummaryList({
+                    rows: licenseDetailsSummaryRows
+                }) }}
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk:59]: Did you create the ticket to refactor this in the way we've done for other summary pages?

**Review by jaucourt — CHANGES_REQUESTED** [review]
Just a couple of final things, thanks for your efforts with this

File: packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk (hunk 7 above)
Comment (ScottDormand96) [reply packages/gafl-webapp-service/src/pages/order-complete/licence-details/licence-details.njk:59]: yeah its: https://eaflood.atlassian.net/browse/IWTF-3153

**Review by jaucourt — APPROVED** [review]

**Pull request merged**